src/PostgREST/MediaType.hs

```diff
@@ -114,48 +118,67 @@ toMimePlanFormat PlanText = "text"
+        ("application", "x-www-form-urlencoded", _) -> MTUrlEncoded
+        ("application", "octet-stream", _)          -> MTOctetStream
+        ("application", "vnd.pgrst.plan", _)        -> getPlan PlanText
+        ("application", "vnd.pgrst.plan+text", _)   -> getPlan PlanText
+        ("application", "vnd.pgrst.plan+json", _)   -> getPlan PlanJSON
+        ("application", "vnd.pgrst.object+json", _) -> MTVndSingularJSON strippedNulls
+        ("application", "vnd.pgrst.object", _)      -> MTVndSingularJSON strippedNulls
+        ("application", "vnd.pgrst.array+json", _)  -> checkArrayNullStrip
+        ("application", "vnd.pgrst.array", _)       -> checkArrayNullStrip
+        ("*","*",_)                                 -> MTAny
+        _                                           -> MTOther mt'
+      where
+        (mainType, subType, params') = tokenizeMediaType mt'
+        params = fromList $ map (first T.toLower) params' -- normalize parameter names to lowercase, per RFC 7321
+        getPlan fmt = MTVndPlan mtFor fmt $
+          [PlanAnalyze  | inOpts "analyze" ] ++
+          [PlanVerbose  | inOpts "verbose" ] ++
+          [PlanSettings | inOpts "settings"] ++
+          [PlanBuffers  | inOpts "buffers" ] ++
+          [PlanWAL      | inOpts "wal"     ]
+          where
+            mtFor = decodeMediaType' $ fromMaybe "application/json" (params !? "for")
+            inOpts str = str `elem` opts
+            opts = T.splitOn "|" $ fromMaybe mempty (params !? "options")
+        strippedNulls = fromMaybe "false" (params !? "nulls") == "stripped"
+        checkArrayNullStrip = if strippedNulls then MTVndArrayJSONStrip else MTApplicationJSON
 
-    checkSingularNullStrip ["nulls=stripped"] = MTVndSingularJSON True
-    checkSingularNullStrip  _                 = MTVndSingularJSON False
-
-    getPlan fmt rest =
-      let
-        opts         = BS.split (BS.c2w '|') $ fromMaybe mempty (BS.stripPrefix "options=" =<< find (BS.isPrefixOf "options=") rest)
-        inOpts str   = str `elem` opts
-        dropAround p = BS.dropWhile p . BS.dropWhileEnd p
-        mtFor        = fromMaybe MTApplicationJSON $ do
-          foundFor    <- find (BS.isPrefixOf "for=") rest
-          strippedFor <- BS.stripPrefix "for=" foundFor
-          pure . decodeMediaType $ dropAround (== BS.c2w '"') strippedFor
-      in
-      MTVndPlan mtFor fmt $
-        [PlanAnalyze  | inOpts "analyze" ] ++
-        [PlanVerbose  | inOpts "verbose" ] ++
-        [PlanSettings | inOpts "settings"] ++
-        [PlanBuffers  | inOpts "buffers" ] ++
-        [PlanWAL      | inOpts "wal"     ]
+-- | Split a Media Type string into components
+-- >>> tokenizeMediaType "application/vnd.pgrst.plan+json;for=\"text/csv\""
+-- ("application","vnd.pgrst.plan+json",[("for","text/csv")])
+-- >>> tokenizeMediaType "*/*"
+-- ("*","*",[])
+tokenizeMediaType :: Text -> (Text, Text, [(Text, Text)])
+tokenizeMediaType t = (mainType, subType, params)
```

Comment:
Doctests were extremely helpful during development, so. 

The proper parser is definitely in order because we do not support quoted parameter names properly; i.e. `application/json;for="this;that" are not supported by the tokenizer.

Created https://github.com/PostgREST/postgrest/issues/3498 so it won't slip between the cracks